ports/esp32/esp32_sdcard.c

```diff
@@ -0,0 +1,386 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2019 Nicko van Someren
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <string.h>
+
+#include "py/runtime.h"
+#include "py/mphal.h"
+#include "extmod/vfs_fat.h"
+
+#include "driver/sdmmc_host.h"
+#include "driver/sdspi_host.h"
+#include "sdmmc_cmd.h"
+#include "esp_log.h"
+
+#define DEBUG 0
+#if DEBUG
+#define DEBUG_printf(...) ESP_LOGI("modsdcard", __VA_ARGS__)
+#else
+#define DEBUG_printf(...) (void)0
+#endif
+
+//
+// There are three layers of abstraction: host, slot and card.
+// Creating an SD Card object will initialise the host and slot.
+// Cards gets initialised by ioctl op==1 and de-inited by ioctl 2
+// Hosts are de-inited in __del__. Slots do not need de-initing.
+//
+
+// Currently the ESP32 Library doesn't support MMC cards, so
+// we don't enable on MICROPY_HW_ENABLE_MMCARD.
+#if MICROPY_HW_ENABLE_SDCARD
+
+// Forward declaration
+const mp_obj_type_t esp32_sdcard_type;
+
+typedef struct _sdcard_obj_t {
+    mp_obj_base_t base;
+    mp_int_t flags;
+    sdmmc_host_t host;
+    // The card structure duplicates the host. It's not clear if we
+    // can avoid this given the way that it is copied.
+    sdmmc_card_t card;
+} sdcard_card_obj_t;
+
+#define SDCARD_CARD_FLAGS_HOST_INIT_DONE 0x01
+#define SDCARD_CARD_FLAGS_CARD_INIT_DONE 0x02
+
+#define _SECTOR_SIZE(self) (self->card.csd.sector_size)
+
+STATIC esp_err_t check_esp_err(esp_err_t code) {
```

Comment:
Ideally all cases here would just raise OSError with the appropriate error code.  Eg `OSError(MP_ENOMEM)` which is distinct from `MemoryError`; same with `OSError(MP_ETIMEDOUT)`.

Reply:
Really? OK. I thought it would be more helpful to the user to give them a real `TimeoutError` that make them look inside an `OSError` to see what the system was complaining about. What's your rationale there? Surely the user cares more about the nature of the problem than the component that caused it.

Reply:
With `TimeoutError`: this exception type is not enabled by default and so it's use is discouraged over `OSError(ETIMEDOUT)` (and it seems esp32 is the only port that enables `TimeoutError`, and even then it never uses it).  `OSError`'s will actually render with a nice string description of the error type if it's a common one (and ETIMEDOUT is common).

Reply:
Fair enough. I have switched this to only return `OSError`s.